### ISSUE TITLE
semanticTokensHandler breaks nvim-lsp

### ISSUE THREAD

**Issue by vrischmann** [zigtools/zls#135]
Hello,

I'm using neovim with [lsp](https://neovim.io/doc/user/lsp.html) and [nvim-lsp](https://github.com/neovim/nvim-lsp) (with [this config](https://gist.github.com/vrischmann/7ede9bff28fe3ad4e4d8fc9fc96e59c7) if you're curious).

As far as I can see neovim doesn't implement `textDocument/semanticTokens` and unfortunately when it receives even a response with no tokens it shows an error:
```
[ ERROR ] 2020-07-14T02:55:14+0200 ] /usr/share/nvim/runtime/lua/vim/lsp.lua:363 ]	"LSP[zls]"	"on_error"	{  code = "NO_RESULT_CALLBACK_FOUND",  err = {    id = 0,    jsonrpc = "2.0",    result = {      data = {}    }  }}
```

That error seems to come [from here](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/rpc.lua#L400-L418) but I'm unsure if there's a bug in neovim or in zls.

Right now to work around this I set `enable_semantic_tokens = false` in the config and apply the following patch:
```
diff --git a/src/main.zig b/src/main.zig
index e95b4a1..822efb8 100644
--- a/src/main.zig
+++ b/src/main.zig
@@ -1170,8 +1170,7 @@ fn semanticTokensHandler(arena: *std.heap.ArenaAllocator, id: types.RequestId, r
             .id = id,
             .result = .{ .SemanticTokens = .{ .data = token_array } },
         });
-    } else
-        return try respondGeneric(id, no_semantic_tokens_response);
+    }
 }

 fn completionHandler(arena: *std.heap.ArenaAllocator, id: types.RequestId, req: requests.Completion, config: Config) !void {
```

which fixes the error.

**Comment by adam-r-kowalski** [comment]
This patch fixed the problem for me as well!

**Comment by alexnask** [comment]
Hm, this is weird, this function should not be called at all if the client does not support semantic token highlighting.  
I think this is a compiler bug that causes us to call invalid handlers in [here](https://github.com/zigtools/zls/blob/master/src/main.zig#L1412-L1440) (which is already ugly code because of compiler bugs :( ).  
I will rewrite this to avoid this bug and test on nvim-lsp.  
Thanks for reporting.

**Comment by alexnask** [comment]
Should be fixed in 9aa11276a8a8523af645d3f161174879721e2e79.  
Thanks for reporting and sorry for taking this long to get back to you.
